rust/connlib/tunnel/src/lib.rs

```diff
@@ -237,27 +220,6 @@ where
     }
 }
 
-#[cfg_attr(not(target_os = "android"), allow(unused_variables))]
-fn new_io<CB>(callbacks: &CB) -> Result<Io>
-where
-    CB: Callbacks,
-{
-    let io = Io::new()?;
-
-    // TODO: Eventually, this should move into the `connlib-client-android` crate.
-    #[cfg(target_os = "android")]
-    {
-        if let Some(ip4_socket) = io.sockets_ref().ip4_socket_fd() {
-            callbacks.protect_file_descriptor(ip4_socket);
-        }
-        if let Some(ip6_socket) = io.sockets_ref().ip6_socket_fd() {
-            callbacks.protect_file_descriptor(ip6_socket);
-        }
```

Comment:
If it's possible to protect the file descriptors from inside connlib and inside the tunnel process, why did it move outside? Is it for error handling?

This is something that's confused me a little. Connlib _can_ be platform-specific, right? Like `tun_linux.rs` and `tun_windows.rs`. We're not just moving things out to clients and across the FFI boundary just because, right?

Reply:
In this case, `protect` can only be called from JVM code so we need a callback.

I think platform-specific callbacks are confusing. I'd rather shuffle the logic around so the individual client glue-code crates can do this.

I'd still call this "connlib being platform specific" because the Android app just uses a `Session`.

One alternative could be to store a dynamic-dispatched callback pointer inside `Sockets`, maybe that is even cleaner? It would be "invisible" from the outside.

Reply:
What do you think of [2c2c617](https://github.com/firezone/firezone/pull/4263/commits/2c2c6178cbfa9b5d6ed7ee69aef3727ba6ebced2)?